Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift

```diff
@@ -0,0 +1,241 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import struct Foundation.Date
+import class Foundation.JSONDecoder
+import struct Foundation.NSRange
+import class Foundation.NSRegularExpression
+import struct Foundation.URL
+import PackageModel
+import TSCBasic
+
+struct GitHubPackageMetadataProvider: PackageMetadataProvider {
+    let httpClient: HTTPClient
+    let defaultHttpClient: Bool
+    let decoder: JSONDecoder
+    let queue: DispatchQueue
+
+    init(httpClient: HTTPClient? = nil) {
+        self.httpClient = httpClient ?? .init()
+        self.defaultHttpClient = httpClient == nil
+        self.decoder = JSONDecoder()
+        #if os(Linux)
+        self.decoder.dateDecodingStrategy = .iso8601
+        #else
+        if #available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *) {
+            self.decoder.dateDecodingStrategy = .iso8601
+        } else {
+            self.decoder.dateDecodingStrategy = .customISO8601
+        }
+        #endif
+        self.queue = DispatchQueue(label: "org.swift.swiftpm.GitHubPackageMetadataProvider", attributes: .concurrent)
+    }
+
+    func get(_ reference: PackageReference, callback: @escaping (Result<PackageCollectionsModel.PackageBasicMetadata, Error>) -> Void) {
```

Comment:
@yim-lee changed the flow here so it first tries to get the repo info, then if successful goes and fans out the rest of API calls, it also throws NotFoundError if repo is not found